tests/test_irdl.py

```diff
@@ -0,0 +1,243 @@
+
+    param1: ParameterDef[Attribute]
+    param2: ParameterDef[Attribute]
+
+
+def test_eq_attr_verify():
+    """Check that an EqAttrConstraint verifies the expected attribute"""
+    bool_true = BoolData(True)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    eq_true_constraint.verify(bool_true)
+
+
+def test_eq_attr_verify_wrong_parameters_fail():
+    """
+    Check that an EqAttrConstraint fails to verify an attribute with different
+    parameters.
+    """
+    bool_true = BoolData(True)
+    bool_false = BoolData(False)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(bool_false)
+    assert e.value.args[0] == (
+        f"Expected attribute {bool_true} but got {bool_false}")
+
+
+def test_eq_attr_verify_wrong_base_fail():
+    """
+    Check that an EqAttrConstraint fails to verify an attribute with a
+    different base attribute.
+    """
+    bool_true = BoolData(True)
+    int_zero = IntData(0)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(int_zero)
+    assert e.value.args[0] == (
+        f"Expected attribute {bool_true} but got {int_zero}")
+
+
+def test_base_attr_verify():
+    """
+    Check that a BaseAttr constraint verifies an attribute with the expected
+    base attribute.
+    """
+    eq_true_constraint = BaseAttr(BoolData)
+    eq_true_constraint.verify(BoolData(True))
+    eq_true_constraint.verify(BoolData(False))
+
+
+def test_base_attr_verify_wrong_base_fail():
+    """
+    Check that a BaseAttr constraint fails to verify an attribute with a
+    different base attribute.
+    """
+    eq_true_constraint = BaseAttr(BoolData)
+    int_zero = IntData(0)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(int_zero)
+    assert e.value.args[0] == (
+        f"{int_zero} should be of base attribute {BoolData.name}")
+
+
+def test_any_attr_verify():
+    """Check that an AnyAttr verifies any attribute."""
+    any_constraint = AnyAttr()
+    any_constraint.verify(BoolData(True))
+    any_constraint.verify(BoolData(False))
+    any_constraint.verify(IntData(0))
+
+
+def test_anyof_verify():
+    """
+    Check that an AnyOf constraint verifies if one of the constraints
+    verify.
+    """
+    constraint = AnyOf([BaseAttr(BoolData), EqAttrConstraint(IntData(0))])
+    constraint.verify(BoolData(True))
+    constraint.verify(BoolData(False))
+    constraint.verify(IntData(0))
+
+
+def test_anyof_verify_fail():
+    """
+    Check that an AnyOf constraint fails to verify if none of the constraints
+    verify.
+    """
+    constraint = AnyOf([BaseAttr(BoolData), EqAttrConstraint(IntData(0))])
+    with pytest.raises(VerifyException) as e:
+        constraint.verify(IntData(1))
+    assert e.value.args[0] == f"Unexpected attribute {IntData(1)}"
+
+
+@dataclass
+class LessThan(AttrConstraint):
```

Comment:
Hmmm, do you mean having `LessThan(BaseAttrConstraint)` instead?
That's an interesting idea, but I'm not completely sure if this is the right idea.
The reason is that users might overwrite `verify` without calling `super`, and in general iRDL (the C++ version) will not inherit from `BaseAttrConstraint`, because it makes code generation and optimization a bit harder.
However, the way we should deal with this, is to use `Annotated[BaseAttr[MyAttr], MyCustomConstraint]`, I can fix this to make it look better.

Reply:
So maybe to express more what I mean here.
Currently, `AttrConstraint` is a class, that we inherit.
However, in IRDL design, this would look more like an ADT, where we have those cases:
* Any
* AnyOf
* AllOf
* Base
* Params
* Custom

Custom is separate from the others, because it is the only non-declarative one that we can't introspect.
In any way, if we would make our custom inherit from `BaseAttr`, then we can't reason about `BaseAttr` as we did before, or at least having the `Base` part and the `Custom` part separate would make this better.
I'm currently thinking of this design in the C++ world, so I'll guess I'll modify how we do this in the Python world once we manage to have a good design!

Reply:
So for now, I'm checking in our custom constraint with an `if` (that will never be executed).

Reply:
I am not sure how familiar you are with Scala mix-ins (or traits), but essentially, these are used to define building blocks which can be mixed into classes later on. My comment comes from me feeling that we are doing something similar here. You have some classes which model constraints on the attributes. I think it would be really cool if you could mix several of these together to get a combined constraint.

Not sure, maybe this is not even possible in Python. However, I feel like such an interface migth be more intuitive than an assertion somewhere in the code.

I am fine with merging as is though :)

Reply:
I understand what you mean!
We could think of using mix-ins for this, though the translation to MLIR IRDL would thus get more complex, and I'm not sure if there would always be a clear translation.

Reply:
Yeah fair, the fact that this needs to work with the MLIR IRDL, which probably uses a lot of C++, makes it more sensible to keep this design. Though, I am wondering whether xDSL might become big enough to varrant such design differences...

Reply:
Yeah, it depends a lot on which direction we will go.
I would say that at least for now, we want to keep an easy MLIR/xDSL conversion, though yes, we might change this at some point!